---
title: "DEG analysis and GO enrichment"
author: "MSK"
date: "2024-01-25"
output:
  html_document:
    df_print: paged
---

## Introduction
This script is used to perform DEG analysis by DESeq2 and to perform GO enrichment using the identified DE genes.

TCGA-LIHC RNAseq data and clinical data have been previously downloaded and processed. DESeq2 is used to identify the genes which are differentially expressed in the patients with HIGH overall survival status (os_status) compared to LOW os_status. 

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning=FALSE)
knitr::opts_chunk$set(fig.width=8, fig.height=12, fig.align = "center") 

# load libraries
# library(TCGAbiolinks)
library(tidyverse)
library(DESeq2)
library(clusterProfiler)
library(org.Hs.eg.db)
library(enrichplot)
library(ggupset)

# BiocManager::install("clusterProfiler")
# BiocManager::install("org.Hs.eg.db")
# # For visualisation
# install.packages('pheatmap')
# install.packages("DOSE")
# install.packages("enrichplot")

# load count data 
lihc.count <- read_tsv(file = 'data/TCGA-LIHC_unstrand_rawcount_369samples.tsv') %>%
  column_to_rownames(var = 'gene_id')

# load sample data 
coldata <- read_tsv(file = 'data/TCGA-LIHC_clinicaldata_369samples.tsv')
# make sure that count data and sample data have same sample names with same order
all(colnames(lihc.count) == coldata$patient) # TRUE

# load gene data 
gene.metadata <- read_tsv(file = 'data/TCGA-LIHC_genemetadata.tsv') 

```

## DESeq2

DESeq2 to identify the genes that are expressed differentially in 2 classes of os_status.

```{r}
## create DESeqDataSet from count matrix using DESeqDataSetFromMatrix ----
# set os_status as factors
coldata$os_status <- factor(coldata$os_status, levels = c('low', 'high'))

# Setting up countData object
dds <- DESeqDataSetFromMatrix(countData = lihc.count,
                              colData = coldata,
                              design = ~ os_status)
## filter out genes with low expression ----
# keep genes with at least 10 reads in at least 75% of samples
keep <- rowSums(counts(dds) >=10) >= 369*0.75
dds <- dds[keep,] # 13264   369

dds
## Run DESeq ----------------------
dds <- DESeq(dds)
res <- results(dds)

```

A quick summary of DESeq2 results with p-value < 0.1 :
```{r}
print(summary(res))
```


```{r} 
# keep genes with absolute value of log2FoldChange > 0.5 with assocaiated pvalue < 0.05 as DE genes
sig.res <- res[(abs(res$log2FoldChange) > 0.5 & res$pvalue < 0.05), ]  %>% as.data.frame() %>%
  rownames_to_column(var = 'gene_id') %>%
  left_join(gene.metadata %>% dplyr::select(gene_id, gene_name)) %>%
  dplyr::select(!gene_id) %>%
  relocate('gene_name', .before = 'baseMean')

sig.df <- sig.res %>% as.data.frame() %>%
  mutate(diffexpressed = case_when(
    log2FoldChange > 0 & pvalue < 0.05 ~ 'UP',
    log2FoldChange < 0 & pvalue < 0.05 ~ 'DOWN',
    pvalue > 0.05 ~ 'NO'
  )) 
# quick look to know how many UP and down regulated genes 
table(sig.df$diffexpressed) 

# Split the dataframe into a list of sub-dataframes: upregulated, downregulated genes
deg_results_list <- split(sig.df, sig.df$diffexpressed)

# Run enrichGO on each sub-dataframe
res <- lapply(names(deg_results_list),
              function(x) enrichGO(gene = deg_results_list[[x]]$gene_name,
                                   OrgDb = "org.Hs.eg.db", keyType = "SYMBOL", ont = "BP")) # entichGO will keep only padj < 0.05
names(res) <- names(deg_results_list)

```
### GO enrichment of up-regulated genes

#### Visualization by barplot 
The top 20 most significant enriched terms have been displayed in the bar plot: several pathways involved in immune response activating / regulation are enriched in High os_status group. 


```{r}
upGO <- res[['UP']]
barplot(upGO, showCategory = 20)
```

#### Visualization by cnetplot

A cnetplot shows the linkages of genes and biological pathways, represented as a network. The red big nodes are pathways to which the genes are associated. This allows to identify clusters of genes that may share common biological functions or participate in the same molecular pathways. For example, ILR7, SLAMF1, RASGRP1, HLA-DOA, CD40LG and CD2 which participate in both ‘lymphocyte mediated immunity’ and ‘positive regulation of cell activation’ pathways.

```{r, fig.width=12, fig.height=12, fig.align = "center"}

cnetplot(upGO, node_label="all",
         color.params = list(category = 'firebrick', gene = 'steelblue'),
         cex.params = list(category_label = 1, gene_label = 0.8)) 

```

#### Visualization by Heatplot 

Similar to cnetplot, the heatplot can simplify the result and more easy to identify expression patterns when there are many significant pathways to visualize. 


```{r, fig.width=12, fig.height=12, fig.align = "center"}
# Heatplot
heatplot(upGO, showCategory = 30)
```

#### Visualization by UpSet Plot 

```{r, fig.width=10, fig.height=8, fig.align = "center"}
upsetplot(upGO)

```

#### Enrichment Map

```{r, fig.width=12, fig.height=12, fig.align = "center"}
enupGO <- pairwise_termsim(upGO)

# emapplot(enupGO, cex_category=1.5)
emapplot(enupGO, layout="kk")

```


## Session Information

```{r}
print(sessionInfo())
```
